lib/grizzly/zwave_firmware.ex

```diff
@@ -0,0 +1,165 @@
+  def maybe_run_zwave_firmware_update(options) do
+    with :ok <- check_zwave_programmer_exists(options),
+         :ok <- maybe_upload_zwave_firmware_update(options) do
+      :ok
+    else
+      {:error, reason} ->
+        Logger.info("[Grizzly] The hub's z-wave firmware was NOT updated: #{inspect(reason)}")
+    end
+
+    :ok
+  end
+
+  defp check_zwave_programmer_exists(options) do
+    if File.exists?(options.zw_programmer_path) do
+      :ok
+    else
+      {:error, :zw_programmer_does_not_exist}
+    end
+  end
+
+  defp maybe_upload_zwave_firmware_update(options) do
+    with {:ok, current_firmware} <- current_firmware(options),
+         {:ok, fw_image_file} <- fw_image_file(options.zwave_firmware, current_firmware) do
+      upload_zwave_firmware_update(fw_image_file, options)
+    else
+      {:error, reason} ->
+        Logger.info("[Grizzly] The hub's zwave firmware is not being updated: #{reason}")
+    end
+
+    :ok
+  end
+
+  # Looks for a valid upgrade fw image file
+  defp fw_image_file(firmware_info, current_firmware) do
+    case applicable_firmware_info(firmware_info, current_firmware) do
+      %{path: path} ->
+        if File.exists?(path) do
+          {:ok, path}
+        else
+          {:error, :fw_image_file_not_found}
+        end
+
+      nil ->
+        Logger.info(
+          "[Grizzly] No applicable firmware upgrade found in #{inspect(firmware_info)} for current firmware #{inspect(current_firmware)}"
+        )
+
+        {:error, :fw_image_file_not_found}
+    end
+  end
+
+  defp applicable_firmware_info(
+         firmware_info,
+         %{chip_type: chip_type, version: current_version} = _current_firmware
+       ) do
+    Enum.find(firmware_info, fn %{
+                                  chip_type: target_chip_type,
+                                  version: target_version,
+                                  path: path
+                                } ->
+      target_chip_type == chip_type and
+        upgrade_valid?(current_version, target_version) and
+        intended_version?(path, target_version)
+    end)
+  end
+
+  # This assumes that the firmware version is in the file's name. Belt and suspenders. Probably will be removed.
+  defp intended_version?(path, target_version) do
+    with {:ok, path_version} <- fw_image_version(path),
+         {:ok, version} <- Version.parse(target_version) do
+      Version.compare(version, path_version) == :eq
+    else
+      _error ->
+        Logger.warn(
+          "[Grizzly] The firmware info's path #{inspect(path)} does not match announced version #{inspect(target_version)}"
+        )
+
+        false
+    end
+  end
+
+  # .../ZW_SerialAPI_Controller_7.16.3_297_ZGM130S_REGION_US_LR.gbl
+  defp fw_image_version(fw_image_file) do
+    basename = Path.basename(fw_image_file)
+    regex = ~r/\w+_(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+).*\.gbl/
+
+    Regex.named_captures(regex, basename)
+    |> parse_version_capture()
+  end
```

Comment:
```suggestion
```

This code can be deleted. It won't work for some firmware files since the naming is different on them, and I don't think it's worth fixing.